### DISCUSSION TITLE
Multiple GameScene? Separate 2D and 3D in different SubScene?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1057] [Ideas]
Hi:

JavaFX provides a way of mix 2D and 3D object together using SubScene
https://openjfx.io/javadoc/16/javafx.graphics/javafx/scene/SubScene.html
so is it possible for FXGL to provide similar mechanism that we could have a sub scene of 3D objects and overlap with 2D scene layer?
thus we could mix 2D rendered objects and 3D objects in the same screen with different scenes
for now we could only choose the objects in the game are all 2D or 3D because they are all in the same scene
is it possible to build them and attach them in different scene layers 
and then overlap these scenes layers and show them in the same screen?
like a back sub scene with all 3D objects and a front default scene with all 2D objects?

Thanks

**Comment by chengenzhao** [comment]
I tried this code but it doesn't work

 ```java
   var obj =spawn("3DObject");
    var subscene = new SubScene(obj.getViewComponent().getParent(), 1000,1000,true, SceneAntialiasing.BALANCED);
    getGameScene().addChild(subscene);
```

exception msg

java.lang.IllegalArgumentException: Group@756d1abbis already inside a scene-graph and cannot be set as root
	at javafx.graphics/javafx.scene.SubScene$2.invalidated(SubScene.java:290)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at javafx.graphics/javafx.scene.SubScene.setRoot(SubScene.java:263)
	at javafx.graphics/javafx.scene.SubScene.<init>(SubScene.java:203)
	at xtrike@1.0/com.whitewoodcity.battlefxgl.Game.initBattleField(Game.java:110)
	at xtrike@1.0/com.whitewoodcity.battlefxgl.BattleFXGLPane$1.call(BattleFXGLPane.java:40)
	at xtrike@1.0/com.whitewoodcity.battlefxgl.BattleFXGLPane$1.call(BattleFXGLPane.java:33)
	at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)

seems like entity is already belonged to the default scene
so is it possible to move 3D objects into a sub scene that is created purely for 3D objects 

Thanks

**Comment by chengenzhao** [comment]
If the scene layers are separated
we could provide different settings for different scenes^_^
like SceneAntialiasing.BALANCED

**Comment by chengenzhao** [comment]
Seems like fxgl-scene project has different types of scene
so would be nice to have a separate 3d based sub scene of fxgl by extending SubScene class of JavaFX?
and allow developers to choose which scene that the entity.viewComponents should be in?

**Comment by chengenzhao** [comment]
I found the source code here
 ```java
           gameScene = GameScene(settings.width, settings.height,
                    GameWorld(),
                    PhysicsWorld(settings.height, settings.pixelsPerMeter, settings.collisionDetectionStrategy),
                    settings.is3D
            )
```

now there is only one game scene and we have to choose is3D or not
so it would be nice to have another game scene like game scene 0 and game scene 1
game scene 0 is 2D //is3D = false
game scene 1 is 3D //is3D = true
but gameScene 0&1 share the same game world and physics world
and the view components are bond to different game scenes
and we could have different settings, e.g. for 3D we could enable anti-aliasing using JavaFX SubScene to implement that
in that case we may need to place a JavaFX SubScene in the JavaFX Pane used to implement FXGL GameScene.

**Comment by chengenzhao** [comment]
It could be implemented by insert a JavaFX SubScene class between root and contentRoot(both types are JavaFX Pane) of FXGL SubScene

<img width="497" alt="F79BE846-D605-4B88-9AB9-0983E3334D11" src="https://user-images.githubusercontent.com/5525436/120055478-f5ecca00-c068-11eb-817d-4ffc292c8bf9.png">

**Comment by AlmasB** [comment]
There is an [open issue](https://github.com/AlmasB/FXGL/issues/995) related to supporting entities in multiple scene types, not just GameScene. After that is supported, it will be possible to properly consider multiple GameScene's, or perhaps a number of arbitrary scenes that allow mixing multiple (2d/3d) modes without changing the current game scene since FXGL thinks there is always just 1 instance.
